### ISSUE TITLE
Pushing a clip on top of clip_and_scroll_info has unexpected behaviour

### ISSUE THREAD

**Issue by staktrace** [servo/webrender#1326]
Apply this patch to examples/scrolling.rs:

```
diff --git a/webrender/examples/scrolling.rs b/webrender/examples/scrolling.rs
index 9abc12e..5ce91a9 100644
--- a/webrender/examples/scrolling.rs
+++ b/webrender/examples/scrolling.rs
@@ -5,19 +5,17 @@
 extern crate gleam;
 extern crate glutin;
 extern crate webrender;
 extern crate webrender_traits;

 use gleam::gl;
 use std::env;
 use std::path::PathBuf;
-use webrender_traits::{ClipId, ColorF, DeviceUintSize, Epoch, LayoutPoint, LayoutRect};
-use webrender_traits::{LayoutSize, PipelineId, ScrollEventPhase, ScrollLocation, TransformStyle};
-use webrender_traits::WorldPoint;
+use webrender_traits::*;

 struct Notifier {
     window_proxy: glutin::WindowProxy,
 }

 impl Notifier {
     fn new(window_proxy: glutin::WindowProxy) -> Notifier {
         Notifier {
@@ -156,29 +154,61 @@ fn main() {
                                                  Some(ClipId::new(43, pipeline_id)));
         builder.push_clip_id(nested_clip_id);
         // give it a giant gray background just to distinguish it and to easily
         // visually identify the nested scrollbox
         let clip = builder.push_clip_region(&(-1000, -1000).to(5000, 5000), vec![], None);
         builder.push_rect((-1000, -1000).to(5000, 5000),
                           clip,
                           ColorF::new(0.5, 0.5, 0.5, 1.0));
-        // add a teal square to visualize the scrolling/clipping behaviour
-        // as you scroll the nested scrollbox with WASD keys
+
+        // let's simulate background-attachment:fixed using a teal square. first
+        // we'll set up a clip ("bgfixed_clip_id") that scrolls around inside
+        // nested_clip_id.
+        let bgfixed_clip = builder.push_clip_region(&(0, 100).to(50, 200), vec![], None);
+        let bgfixed_clip_id = builder.define_clip((0, 100).to(50, 200),
+                                                  bgfixed_clip,
+                                                  None);
+        builder.push_clip_id(bgfixed_clip_id);
+
+        // now we push a ClipAndScrollInfo that makes stuff scroll with clip_id
+        // but be clipped by bgfixed_clip_id. Since clip_id is the ancestor
+        // scroll layer of nested_clip_id, this effectively makes the contents
+        // fixed with respect to nested_clip_id, but still be subjected to the
+        // scrolling clip as nested_clip_id is scrolled around.
+        builder.push_clip_and_scroll_info(ClipAndScrollInfo::new(clip_id, bgfixed_clip_id));
+
+        // Adding this extra clip (which is exactly the same as the clip on the
+        // teal square below) breaks the background-attachment:fixed effect.
+// !    let clip = builder.push_clip_region(&(0, 100).to(50, 150), vec![], None);
+// !    let redundant_clip_id = builder.define_clip((0, 100).to(50,150),
+// !                                                clip,
+// !                                                None);
+// !    builder.push_clip_id(redundant_clip_id);
+
+        // this is a teal square that is the "content" that behaves like
+        // background-attachment:fixed
         let clip = builder.push_clip_region(&(0, 100).to(50, 150), vec![], None);
         builder.push_rect((0, 100).to(50, 150),
                           clip,
                           ColorF::new(0.0, 1.0, 1.0, 1.0));
+
+// !    builder.pop_clip_id(); // redundant_clip_id
+        builder.pop_clip_id(); // clip_and_scroll_info
+
+        builder.pop_clip_id(); // bgfixed_clip_id
+
         // just for good measure add another teal square in the bottom-right
         // corner of the nested scrollframe content, which can be scrolled into
         // view by the user
         let clip = builder.push_clip_region(&(250, 350).to(300, 400), vec![], None);
         builder.push_rect((250, 350).to(300, 400),
                           clip,
                           ColorF::new(0.0, 1.0, 1.0, 1.0));
+
         builder.pop_clip_id(); // nested_clip_id

         builder.pop_clip_id(); // clip_id
         builder.pop_stacking_context();
     }

     builder.pop_stacking_context();

```

Run the example and scroll the nested scrollframe. Observe how the teal square is clipped as you scroll, behaving the way you'd expect background-attachment:fixed to behave.

Now uncomment the lines that start with "// !" and repeat. Now the teal square doesn't get clipped as you scroll.

It seems that pushing an extra clip after pushing the clip_and_scroll_info doesn't *combine* the clips like I would expect. Instead the scrolling clip appears to be *replaced* by the new clip that is pushed, and so we lose the scrolling clip and the teal square is no longer clipped as you scroll the nested scrollframe.

/cc @mrobinson

**Comment by mrobinson** [comment]
@staktrace Sorry for the delay in replying. I think what is happening here is that `redundant_clip_id` is being positioned relative to `clip_id` instead of `nested_clip_id`. The key thing here is that a clip is placed into the clip tree based on the current positioning clip_id on the stack at the moment you call `define_clip`. Take a look at this code from the `DisplayListBuilder`:

```rust
    pub fn define_clip(&mut self,
                       content_rect: LayoutRect,
                       _token: ClipRegionToken,
                       id: Option<ClipId>)
                       -> ClipId {
        let id = match id {
            Some(id) => id,
            None => {
                self.next_clip_id += 1;
                ClipId::Clip(self.next_clip_id - 1, self.pipeline_id)
            }
        };

        let item = SpecificDisplayItem::Clip(ClipDisplayItem {
            id: id,
            parent_id: self.clip_stack.last().unwrap().scroll_node_id,
        });

        self.push_item(item, content_rect);
        id
    }
``` 

So what's happening here is that you do:
```rust
builder.push_clip_and_scroll_info(ClipAndScrollInfo::new(clip_id, bgfixed_clip_id));
```
and then define the redundant clip. If you define the clip before pushing the `ClipAndScrollInfo` and then push both the `ClipAndScrollInfo` and the redundant clip, you should see the behavior you expect.

I think maybe the clip stack makes things a bit too complicated, which is why I'm considering removing it. Instead, it might make more sense to always pass in the parent clip when calling `define_clip`, which should make it clearer what is going to happen inside of WebRender.

**Comment by staktrace** [comment]
I experimented some more and I got the desired effect by doing what you suggested with a slight modification. I had to define redundant_clip_id before pushing the ClipAndScrollInfo. And then I had to push another ClipAndScrollInfo, this time with (clip_id, redundant_clip_id), after pushing the first ClipAndScrollInfo. If I just pushed the redundant_clip_id as a simple clip it didn't work. Which in a way makes sense given how the clipping tree seems to work. I'll have to see if I can make the layer tree in gecko produce this series of calls.
